### DISCUSSION TITLE
Trouble calling named procedures from a host application

### DISCUSSION THREAD

**Discussion by bcalco** [leppie/IronScheme#82] [Discussions]
@leppie 

Inspired by the wiki page describing the IronScheme Extension Methods, specifically this bit:

```csharp
var myproc = "(lambda (x y) (+ x y))".Eval();
var r3 = "({0} {1} {2})".Eval(myproc, 1 , 2); // r3 is 3

"(define foo 1000)".Eval();  // executes definition in interaction environment
var r4 = "foo".Eval(); // r4 is 1000
```

I am trying to support user-defined functions in my WinForms GUI application by allowing users to define lambdas in various contexts. In some contexts the name of the lambda will be implicit and computable, to be generated by my host application. In these cases, the automatic names like "anon$1" won't work for me--I want to have some control over this. So I tried several variations but this was one I think you can reproduce:

```csharp
string lambdaExpression = "(lambda (n) (* n 5))"; // user defined lambda spec
string lambdaInternalName = "foo$bar$baz"; // names will look something like this
string defineStatement = "(define " + lambdaInternalName + " {0})";
defineStatement.Eval (lambdaExpression); // "(define foo$bar$baz (lambda (n) (* n 5)))";
// get passedParameters array, defined as object [], with 1 parameter of value 5 (an int32)
string finalExpression = "(apply " + lambdaInternalName + "  (vector->list {0}))";
var computationResult = finalExpression.Eval ( passedParameters );
// process computationResult further
```

I'm getting "expected a procedure" error. Based on the examples on the wiki page, I was quite hopeful that would work, to "de-anonymize" lambdas and predictably be able to call them in the interactive environment. I'm not quite sure how within C# to inspect or alter that environment except via the Eval functions, but either the symbol foo$bar$baz is not resolved to a procedure, or there is something else I must do to coerce the symbol to act like one. At some point, with some variation of this, I got a different exception that I was trying to access an undefined symbol; but at this point it seems to be that it is recognizing the symbol, just not as a procedure.

In general I'm quite eager to get this resolved so any help appreciated! Additionally, any pointers about memory and environment management from the host-side would be great, as these custom functions will be continually updated.

Thanks in advance, anyone, but especially @leppie or anybody who has a lot of experience interacting with IronScheme from the host side. :)

**Comment by bcalco** [comment]
OK, an update. Since my post I figured maybe the issue was the declaration statement which may have been defining `foo$bar$baz` as a string based on the interpolation of the lambda text. So I got a bit further by doing this:

```csharp
string lambdaExpression = "(lambda (n) (* n 5))"; // user defined lambda spec
string lambdaInternalName = "foo$bar$baz"; // names will look something like this
string defineStatement = "(define " + lambdaInternalName + " " + lambdaExpression +")";
defineStatement.Eval (); // "(define foo$bar$baz (lambda (n) (* n 5)))";
// get passedParameters array, defined as object [], with 1 parameter of value 5 (an int32)
string finalExpression = "(apply " + lambdaInternalName + "  (vector->list {0}))";
var computationResult = finalExpression.Eval ( passedParameters );
// process computationResult further
```

Now the problem was, a lambda with arity 1 doesn't like the `(vector->list ..)` call and was complaining, "not vector".  So I said maybe there's log in there for a single-element object array to be treated like a single value, so I changed the `finalExpression` to:

```csharp
string finalExpression = "(apply " + lambdaInternalName + "  (list {0}))";
```

... and it ran honky-dory. Till I tested a lambda with 2 parameters... So I added logic like this:

```csharp
if ( passedParams.Length == 1 )
{
  finalExpression = "(apply " + lambdaInternalName + " (list {0}))";
}
else
{
  finalExpression = "(apply " + lambdaInternalName + " (vector->list {0}))";
}
```

Now lambdas with 1 parameter still work, but with 2 parameters, say:

```csharp
string lambdaExpression = "(lambda (a b) (* a b))";
...
// passedParams is an object [] containing values { 5, 5}, so expected answer is 25 
```

I get the following un-intuitive error: "Value cannot be null. Parameter name: the name must be a string".

Going to try to call it differently, maybe `lambdaInternalName.Eval (passedParameters);` ... but I feel like I tried that before. However, now that the symbol `foo$bar$baz` is recognized as a procedure perhaps I will have more luck.

**Comment by bcalco** [comment]
Nope, no love. `lambdaInternalName.Eval (passedParameters);` gives me the same "Value cannot be null. Parameter name: the name must be a string" error.

So now I'm not sure what to try next. I'll keep experimenting but now I'm officially winging it, probably having to spelunk in the IronScheme code. Any suggestions gleefully accepted. :)

**Comment by bcalco** [comment]
I am studying `RuntimeExtensions.cs` and getting a sense how to get at the environment and interact with it, and also what is going on to support an interactive environment. I gather it's doing a bit of code generating and the expressions that get executed look rather more complicated than what I think I'm passing in. At least I have an analytical starting point, but how to hook this code up so I can step through it in my project would be appreciated. I may have to copy the IronScheme projects into my solution, maybe as a submodule or something.

**Comment by leppie** [comment]
Here is something to get you started. Does not cover interaction environments.

Start new project, extract IronScheme release zip somewhere, add reference to IronScheme.dll

```
int MySquare(int x) => x * x;

var i = "1".Eval(); // make sure it works
"(library-path (cons \"IronScheme/lib\" (library-path)))".Eval(); // add location of lib files if not default
"(import (ironscheme linq))".Eval(); // import a lib
var r = "(sum (from i in '(1 2 3) select (* i i)))".Eval(); // run something from it

var del = new Func<int, int>(MySquare); 
var proc = del.ToSchemeProcedure(); // define your C# proc to ba callable from Scheme
var rd = "(sum (from i in '(1 2 3) select ({0} i)))".Eval(proc); // just pass it in

var sump = "(lambda x (sum (from i in x select ({0} i))))".Eval<Callable>(proc); // define proc in scheme
var rs = sump.Call(1, 2, 3);
var name = "(procedure-name {0})".Eval(sump); // is anon

var sump2 = "(begin (define (sump2 . x) (sum (from i in x select ({0} i)))) sump2)".Eval<Callable>(proc); // define a define in scheme
var rs2 = sump2.Call(1, 2, 3);
var name2 = "(procedure-name {0})".Eval(sump2); // now has the name sump2

var sumpr = "sump2".Eval<Callable>(); // get back later
var rr = sumpr.Call(1, 2, 3);
var namer = "(procedure-name {0})".Eval(sumpr); // the name is sump2
```

This was all I need to add to project file:
```
  <ItemGroup>
    <Reference Include="IronScheme">
      <HintPath>IronScheme\IronScheme.dll</HintPath>
    </Reference>
  </ItemGroup>

  <ItemGroup>
    <None Update="IronScheme\lib\**\*.*" >
      <CopyToOutputDirectory >PreserveNewest</CopyToOutputDirectory>
    </None>
  </ItemGroup>
```

**Comment by bcalco** [comment]
@leppie - very helpful, thanks!

I got around my immediate problem by creating my own form/expression generator, taking control of the parameters. In the environment I am working with on the C# side, there are some interesting type conversion issues, and so taking control of the code genning vs relying on IronScheme's built in logic to deal with that was probably the root of the problem, though I suspect there is a real bug there in how that IronScheme logic was working. Passing in an object array (i.e., vector) like I was should "just work" but in any case at least now I can debug the code being generated.

Seeing that the `RuntimeExtensions` methods were doing a bit of code genning, I resolved to make sure anything I sent Eval was happy-path, which means not passing parameters and assuming it would do what I needed.

I'm getting a better sense of how the interactive environment works, but I think the DLR interface somewhat obscures my entry into what's really happening, short of adding the IronScheme project to my solution as a submodule or something.

**Comment by bcalco** [comment]
BTW, @leppie, at some point I remember getting an `UnspecifiedObject` back and struggling to know how to deal with it from a type conversion point of view.  Any pointers if that is what is returned from a call to `Eval()`? I would like to do things like call `(interactive-environment-symbols)` and other built in functions for introspection of the environment and know how to use them on the host side, as well.

**Comment by leppie** [comment]
> at some point I remember getting an `UnspecifiedObject`

That is just the `void` instance. You throw it away. Like void return method in C#.

You can look here at some introspection stuff https://github.com/IronScheme/IronScheme/blob/master/IronScheme/IronScheme.Console/psyntax/expander.sls#L3839

**Comment by leppie** [comment]
Here is some more advanced introspection for the (not so advanced) VS plugin.

https://github.com/IronScheme/IronScheme.VisualStudio2/blob/master/IronScheme.VisualStudio/visualstudio.sls

**Comment by leppie** [comment]
Some fun with interaction environment
```
> (define b 42)
> b
42
> (interaction-environment)
#<interactive-environment>
> (interaction-environment (new-interaction-environment))
> b
Unhandled exception during evaluation:
&undefined
&message: "attempted to use undefined symbol"
&irritants: (b)

> (interaction-environment (environment))
> (define a 4)
Unhandled exception during evaluation:
&who: begin
&message: "unbound identifier"
&undefined
&trace: #<syntax begin>

>
```
Now I am stuck! the environment is empty.

From C# you can use say `(new-interaction-environment)` for every user so everyone has a fresh environment (in a server environment). Or you can share 1 between many.
